x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adaptiveallocations/AdaptiveAllocationsScalerService.java

```diff
@@ -0,0 +1,332 @@
+    private final Map<String, Long> lastScaleUpTimesMillis;
+
+    private volatile Scheduler.Cancellable cancellable;
+    private final AtomicBoolean busy;
+
+    public AdaptiveAllocationsScalerService(
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        Client client,
+        SystemAuditor systemAuditor,
+        boolean isNlpEnabled
+    ) {
+        this(threadPool, clusterService, client, systemAuditor, isNlpEnabled, DEFAULT_TIME_INTERVAL_SECONDS);
+    }
+
+    // visible for testing
+    AdaptiveAllocationsScalerService(
+        ThreadPool threadPool,
+        ClusterService clusterService,
+        Client client,
+        SystemAuditor systemAuditor,
+        boolean isNlpEnabled,
+        int timeIntervalSeconds
+    ) {
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+        this.client = client;
+        this.systemAuditor = systemAuditor;
+        this.isNlpEnabled = isNlpEnabled;
+        this.timeIntervalSeconds = timeIntervalSeconds;
+
+        lastInferenceStatsByDeploymentAndNode = new HashMap<>();
+        lastInferenceStatsTimestampMillis = null;
+        lastScaleUpTimesMillis = new HashMap<>();
+        scalers = new HashMap<>();
+        busy = new AtomicBoolean(false);
+    }
+
+    public synchronized void start() {
+        updateAutoscalers(clusterService.state());
+        clusterService.addListener(this);
+        if (scalers.isEmpty() == false) {
+            startScheduling();
+        }
+    }
+
+    public synchronized void stop() {
+        stopScheduling();
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        updateAutoscalers(event.state());
+        if (scalers.isEmpty() == false) {
+            startScheduling();
+        } else {
+            stopScheduling();
+        }
+    }
+
+    private synchronized void updateAutoscalers(ClusterState state) {
+        if (isNlpEnabled == false) {
+            return;
+        }
+
+        TrainedModelAssignmentMetadata assignments = TrainedModelAssignmentMetadata.fromState(state);
+        for (TrainedModelAssignment assignment : assignments.allAssignments().values()) {
+            if (assignment.getAdaptiveAllocationsSettings() != null && assignment.getAdaptiveAllocationsSettings().getEnabled()) {
+                AdaptiveAllocationsScaler adaptiveAllocationsScaler = scalers.computeIfAbsent(
+                    assignment.getDeploymentId(),
+                    key -> new AdaptiveAllocationsScaler(assignment.getDeploymentId(), assignment.totalTargetAllocations())
+                );
+                adaptiveAllocationsScaler.setMinMaxNumberOfAllocations(
+                    assignment.getAdaptiveAllocationsSettings().getMinNumberOfAllocations(),
+                    assignment.getAdaptiveAllocationsSettings().getMaxNumberOfAllocations()
+                );
+            } else {
+                scalers.remove(assignment.getDeploymentId());
```

Comment:
If the model deployment is stopped the assignment will disappear in which case the scaler should be removed. Maybe create a set of used deployment Ids in this for loop then compare it to `scalers.keyset()` and remove the difference.

Reply:
fixed